Fineract 614: Rates module

angelboxes opened the pull request:
Previously defined sub-rates can be added to set the nominal interest rate of a loan product. When a loan account is created you can choose which rates from the associated rates to the loan product should be applied to the new loan account. This is an optional way of defining the nominal rate for loans.

Commit:
A new rates module was added to define new rates that can be used to set min and max nominal interest rate, when a new loan account is created they can be used to determine which rates may be applicable for the loan account.

ShruthiRajaram commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/serialization/JsonParserHelper.java:
```diff
@@ -454,7 +454,7 @@ public static LocalDateTime convertDateTimeFrom(final String dateTimeAsString, f
         if (StringUtils.isNotBlank(dateTimeAsString)) {
             try {
-                eventLocalDateTime = DateTimeFormat.forPattern(dateTimeFormat).withLocale(clientApplicationLocale)
+            eventLocalDateTime = DateTimeFormat.forPattern(dateTimeFormat).withLocale(clientApplicationLocale)
```
Since no changes are made in this file, kindly remove this file from PR

angelboxes replied:
Ok

ShruthiRajaram commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResource.java:
```diff
@@ -178,30 +180,30 @@
+        final LoanScheduleHistoryReadPlatformService loanScheduleHistoryReadPlatformService,
+        final AccountDetailsReadPlatformService accountDetailsReadPlatformService,
+        final EntityDatatableChecksReadService entityDatatableChecksReadService, RateReadService rateReadService,
+        final BulkImportWorkbookService bulkImportWorkbookService,
```
Kindly make RateReadService rateReadService as the last parameter of the method

angelboxes replied:
Ok.

ShruthiRajaram commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/Loan.java:
```diff
@@ -392,20 +395,24 @@
     @OneToOne(cascade = CascadeType.ALL, mappedBy = "loan", optional = true, orphanRemoval = true, fetch=FetchType.EAGER)
     private LoanTopupDetails loanTopupDetails;
 
+    @ManyToMany(fetch = FetchType.EAGER)
```
I think it is OneToMany relationship !!
it possible to change fetchtype to lazy and achieve the expected results?

angelboxes replied:
About this change I think that ManyToMany is okay as we use a third table to join the loan and rate tables just like in the LoanProduct class. I'll make the change and test if the same result can be achieved.

ShruthiRajaram commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanUtilService.java:
```diff
@@ -325,4 +332,35 @@ private LocalDate generateCalculatedRepaymentStartDate(final CalendarHistoryData
+                    rates.add(rate);
+                }
+            }
+        }else{
```
remove else block

angelboxes replied:
removed

ShruthiRajaram commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanUtilService.java:
```diff
@@ -325,4 +332,35 @@ private LocalDate generateCalculatedRepaymentStartDate(final CalendarHistoryData
+        return rates;
+    }
+
+    public Rate findRateByIdIfProvided(final Long rateId) {
```
kindly provide appropriate name

angelboxes replied:
Ok. The name is based on the names used by the methods in the class LoanAssembler. We were just following what we assumed it was a name pattern.

ShruthiRajaram commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanUtilService.java:
```diff
@@ -325,4 +332,35 @@ private LocalDate generateCalculatedRepaymentStartDate(final CalendarHistoryData
+    public Rate findRateByIdIfProvided(final Long rateId) {
+        Rate rate = null;
+        if (rateId != null) {
+            rate = this.rateRepository.findOne(rateId);
```
As per the fineract existing code its advised to access repository methods using wrapper class

angelboxes replied:
I'll use a wrapper class, I just noticed that it is used for the Loan repository but there are some others that don't make use of it so it was a bit misleading.

ShruthiRajaram commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProduct.java:
```diff
@@ -102,6 +103,10 @@
     private List<Charge> charges;
 
+    @ManyToMany(fetch = FetchType.EAGER)
+    @JoinTable(name = "m_product_loan_rate", joinColumns = @JoinColumn(name = "product_loan_id"), inverseJoinColumns = @JoinColumn(name = "rate_id"))
```
@ManyToMany or OneToMany
Can the functionality be achieved by using fetch type lazy

angelboxes replied:
It should be ManyToMany as we use a third table to join the rate and loan_product tables in the same way that the charges and loan_products are joined.  About the fetch type Lazy I guess it should work the same but I will do some tests to be sure.

ShruthiRajaram commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/service/LoanProductReadPlatformServiceImpl.java:
```diff
@@ -465,7 +473,7 @@ public LoanProductData mapRow(final ResultSet rs, @SuppressWarnings("unused") fi
                     floatingRateName, interestRateDifferential, minDifferentialLendingRate, defaultDifferentialLendingRate,
                     maxDifferentialLendingRate, isFloatingInterestRateCalculationAllowed, isVariableIntallmentsAllowed, minimumGap,
-                    maximumGap, syncExpectedWithDisbursementDate, canUseForTopup, isEqualAmortization);
+                    maximumGap, syncExpectedWithDisbursementDate, canUseForTopup, isEqualAmortization, null, this.rates);
```
instead of directly passing null, initialize to a variable (variable name similar to the one in called method ) and pass it, for readability. And follow similar approach in all other places

angelboxes replied:
Ok

ShruthiRajaram commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/service/LoanProductWritePlatformServiceJpaRepositoryImpl.java:
```diff
@@ -304,6 +319,32 @@ private boolean anyChangeInCriticalFloatingRateLinkedParams(JsonCommand command,
+                    if (jsonObject.has("id")) {
+                        final Long id = jsonObject.get("id").getAsLong();
+
+                        final Rate rate = this.rateRepository.findOne(id);
```
Instead of repeated calls to DB, can you try using single query with in clause and validate

angelboxes replied:
Ok

ShruthiRajaram commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/rate/handler/DeleteRateCommandHandler.java:
```diff
@@ -0,0 +1,26 @@
+/**
```
Is this required?

angelboxes replied:
It is not, consider it as gone.

ShruthiRajaram commented on fineract-provider/src/main/resources/sql/migrations/core_db/V343__rates.sql:
```diff
@@ -0,0 +1,62 @@
+  KEY `FK_M_RATE_APPROVE_USER` (`approve_user`),
+  CONSTRAINT `FK_M_RATE_APPROVE_USER` FOREIGN KEY (`approve_user`) REFERENCES `m_appuser` (`id`),
+  CONSTRAINT `FK_M_RATE_CREATE_USER` FOREIGN KEY (`createdby_id`) REFERENCES `m_appuser` (`id`)
+) ENGINE=InnoDB AUTO_INCREMENT=21 DEFAULT CHARSET=latin1;
```
remove ENGINE=InnoDB AUTO_INCREMENT=21 DEFAULT CHARSET=latin1;

angelboxes replied:
Some of the scripts included this part and we guessed it was needed. We'll remove all of them in this new script.

ShruthiRajaram commented on fineract-provider/src/main/resources/sql/migrations/core_db/V343__rates.sql:
```diff
@@ -0,0 +1,62 @@
+  KEY `FK_M_LOAN_RATE_RATE` (`rate_id`),
+  CONSTRAINT `FK_M_LOAN_RATE_LOAN` FOREIGN KEY (`loan_id`) REFERENCES `m_loan` (`id`),
+  CONSTRAINT `FK_M_LOAN_RATE_RATE` FOREIGN KEY (`rate_id`) REFERENCES `m_rate` (`id`)
+) ENGINE=InnoDB DEFAULT CHARSET=latin1;
```
remove ENGINE=InnoDB DEFAULT CHARSET=latin1;

ShruthiRajaram commented on fineract-provider/src/main/resources/sql/migrations/core_db/V343__rates.sql:
```diff
@@ -0,0 +1,62 @@
+  KEY `FK_M_PRODUCT_LOAN_RATE_RATE` (`rate_id`),
+  CONSTRAINT `FK_M_PRODUCT_LOAN_RATE_LOAN` FOREIGN KEY (`product_loan_id`) REFERENCES `m_product_loan` (`id`),
+  CONSTRAINT `FK_M_PRODUCT_LOAN_RATE_RATE` FOREIGN KEY (`rate_id`) REFERENCES `m_rate` (`id`)
+) ENGINE=InnoDB DEFAULT CHARSET=latin1;
```
remove  ENGINE=InnoDB DEFAULT CHARSET=latin1;

ShruthiRajaram commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanUtilService.java:
```diff
@@ -135,10 +135,10 @@ public ScheduleGeneratorDTO buildScheduleGeneratorDTO(final Loan loan, final Loc
         if(isSkipRepaymentOnFirstMonthEnabled){
             isSkipRepaymentOnFirstMonth = isLoanRepaymentsSyncWithMeeting(loan.group(), calendar);
-            if(isSkipRepaymentOnFirstMonth) { numberOfDays = configurationDomainService.retreivePeroidInNumberOfDaysForSkipMeetingDate().intValue(); } 
+            if(isSkipRepaymentOnFirstMonth) { numberOfDays = configurationDomainService.retreivePeroidInNumberOfDaysForSkipMeetingDate().intValue(); }
```
No changes are made to this file. Kindly remove this file from PR

ShruthiRajaram commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/service/LoanProductWritePlatformServiceJpaRepositoryImpl.java:
```diff
@@ -304,6 +318,29 @@ private boolean anyChangeInCriticalFloatingRateLinkedParams(JsonCommand command,
+                        idList.add(jsonObject.get("id").getAsLong());
+                    }
+                }
+                System.out.println(idList.toString());
```
remove Print statement

angelboxes replied:
Ok

ShruthiRajaram commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/service/LoanProductWritePlatformServiceJpaRepositoryImpl.java:
```diff
@@ -304,6 +318,29 @@ private boolean anyChangeInCriticalFloatingRateLinkedParams(JsonCommand command,
+                }
+                System.out.println(idList.toString());
+                rates.addAll(this.rateRepository.findMultipleWithNotFoundDetection(idList));
+                System.out.println(rates.toString());
```
remove Print statement

angelboxes replied:
Ok

ShruthiRajaram commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/rate/domain/RateRepositoryWrapper.java:
```diff
@@ -0,0 +1,72 @@
+    if (rateIds != null && !rateIds.isEmpty()) {
+      final List<Rate> foundRates = this.repository.findAll(rateIds);
+      System.out.println("ids "+rateIds.toString());
+      System.out.println("found "+foundRates.toString());
```
remove Print statement

angelboxes replied:
Ok

ShruthiRajaram commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/rate/serialization/RateDefinitionCommandFromApiJsonDeserializer.java:
```diff
@@ -0,0 +1,128 @@
+
+    final JsonElement element = this.fromApiJsonHelper.parse(json);
+
+    final String name = this.fromApiJsonHelper.extractStringNamed("name", element);
```
Its advised to keep all parameter names in a constant file and use

angelboxes replied:
Ok

vorburger commented:
@angelboxes thank you very much for your interest in and contribution to Fineract/Mifos!  I'm an (old) volunteer who got reminded of Mifos/Fineract this week-end at the FOSDEM conference, and I'm taking a moment to help the project clean up some of it's old issues and pull requests etc.  

We were wondering whether you are still alive, and would be willing to `git rebase develop` this PR and take any previous review feedback into account?  If we don't hear from you within 2 weeks, we will take the liberty to close this PR without merging it, to avoid confusion for future new contributors looking at this repository - hope you understand and that's fair.

/Cc FYI @apache/fineract-committers

angelboxes commented:
> @angelboxes thank you very much for your interest in and contribution to Fineract/Mifos! I'm an (old) volunteer who got reminded of Mifos/Fineract this week-end at the FOSDEM conference, and I'm taking a moment to help the project clean up some of it's old issues and pull requests etc.
> 
> We were wondering whether you are still alive, and would be willing to `git rebase develop` this PR and take any previous review feedback into account? If we don't hear from you within 2 weeks, we will take the liberty to close this PR without merging it, to avoid confusion for future new contributors looking at this repository - hope you understand and that's fair.
> 
> /Cc FYI @apache/fineract-committers

Hi, I pulled the latest changes from the dev repository and rebased my branch.

vorburger commented:
@ShruthiRajaram will you re-review this and merge if OK for you? (From a quick look I've just had, it seems to me that @angelboxes has addressed your review comments, but you'll want to have a closer look I guess.)

vorburger commented:
@angelboxes would you mind to re-rebase this, so that Travis CI will run on it?

angelboxes commented:
> @angelboxes would you mind to re-rebase this, so that Travis CI will run on it?

I just rebased my brach with the latest changes from develop

vorburger commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/rate/domain/Rate.java:
```diff
@@ -0,0 +1,209 @@
+  }
+
+  @Override
+  public int hashCode() {
```
it is almost always a bug to only override `hashCode` and not `equals` ... Google it for more background. May I suggest that you either remove this `hashCode` method or also implement `equals`? I would prefer this not getting merged before that is addressed. Sorry I did not notice this before. PS: https://issues.apache.org/jira/browse/FINERACT-702 - would you have any interest in helping to contribute to that?

angelboxes replied:
Hi, I removed the method as suggested.

vorburger commented:
@angelboxes did you notice that the (new) Travis CI build failed? Note:

`Building 95% > :integrationTest.googlecode.flyway.core.api.FlywayException: Found more than one migration with version '350' (Offenders: SQL '/home/travis/build/apache/fineract/build/tmp/tomcatRunWar/work/Tomcat/localhost/fineract-provider/WEB-INF/classes/sql/migrations/core_db/V350__rates.sql' and SQL '/home/travis/build/apache/fineract/build/tmp/tomcatRunWar/work/Tomcat/localhost/fineract-provider/WEB-INF/classes/sql/migrations/core_db/V350__email_from.sql');`

vorburger reviewed (changes requested):
SQL script needs to be re-numbered

angelboxes commented:
> SQL script needs to be re-numbered

Hi, I didn't noticed there was a more recent Travis build, I've changed the name of the script and rebased my branch. I will keep an eye on the travis build result if it fails again.

vorburger reviewed (approved):
OK for me now! @ShruthiRajaram ok for you as well as-is - merge?

ShruthiRajaram commented:
> OK for me now! @ShruthiRajaram ok for you as well as-is - merge?

ShruthiRajaram commented:
@vorburger , we had a call with this team last year. And we requested some change in functionality. So I prefer Ed Cable or Santosh Math to review the functionality before merging it. I have informed Santosh already and once they confirm we can merge this

vorburger commented:
@ShruthiRajaram my question was just if the points which YOU (earlier, above) raised are now adressed? Voice Calls without any public written record here and no concrete concrete code review feedback which the contributor could react to and rectify, and mention of (nice!) other people who have not commented on this PR in 7 months (!) isn't really how the open source is supposed to work... ;-) Unless you (@ShruthiRajaram) have strong objections, I am inclined to merge this. My main goal isn't this particular functionality, but wanted to encourage contributors like @angelboxes who clearly are reactive to concentrate feedback here - we should not discourage and demotivate such contributions by taking SOOO long...

awasum commented:
Hey @vishwasbabu , Please assign @angelboxes to this issue on JIRA: https://issues.apache.org/jira/projects/FINERACT/issues/FINERACT-614

He is the reporter on the issue.
I tried doing that but for some reason It was failing. Thanks

santoshconflux commented:
Hi @angelboxes, Can we have one quick functionality demo regarding the ticket?

vishwasbabu commented:
I am unable to select him from the typeahead . I shall reach out to infra
for help. Meanwhile , can @angelboxes try assigning the issue to himself ?

On Friday, March 8, 2019, Awasum Yannick <notifications@github.com> wrote:

> Hey @vishwasbabu <https://github.com/vishwasbabu> , Please assign
> @angelboxes <https://github.com/angelboxes> to this issue on JIRA:
> https://issues.apache.org/jira/projects/FINERACT/issues/FINERACT-614
>
> He is the reporter on the issue.
> I tried doing that but for some reason It was failing. Thanks
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/fineract/pull/465#issuecomment-470847270>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAhH6YLI6j6d_KaW6Cdh6n6-El-gZhESks5vUh9agaJpZM4VnVLM>
> .
>


-- 
Sent from my iPad

awasum commented:
That will be a good way to fix it. @angelboxes , please assign the issue to yourself on JIRA.
@angelboxes I appreciate your patience on this issue, you are doing great work.

awasum commented:
@santoshconflux Why do you need a call or demo? Will pulling this code in and testing the feature on your environment not validate if the issue has been solved? 
Also if there needs to be a demo, does it mean a UI is needed because that will mean we accepting this PR only if there is a UI for it already given Fineract does not have a UI. Do you want him to show you a postman call to an endpoint so you see the response. It looks like the issue is not well understood. 
If that is the case, then we need to create a confluence page or this or update the JIRA ticket and even discuss on the mailing list. I think we can merge this code and do the UI tests and confluence pages later given @angelboxes has been so patient with us.

santoshconflux commented:
@awasum , We know @angelboxes did great job including this feature.   Since  UI is not yet developed in Community app level and we still have open ticket https://github.com/openMF/community-app/issues/2997  which GSOC student can work upon. 
All we can check it with Postman only. 
@angelboxes can walk me through Postman or share some use cases. We had demo long time back and in order to just refresh our memory I asked for quick demo , not detailed one, so that I can figure out whether it could create any regressions. 
We should not forget that loan 'product' is very base of whole loan module and any new inclusions in the loan product demands to test whole or at least to some extent of loan module thoroughly for regressions.

santoshconflux commented:
@vishwasbabu , @ShruthiRajaram , if this PR is merged and since it's UI version is not yet implemented, https://github.com/openMF/community-app/issues/2997
Do you think it won't be any problem for existing functionality  if I fetch latest code from develop branch of Community App to work with latest code from develop branch of Fineract?

vishwasbabu commented:
Santosh ,  I haven’t reviewed the code .
-> The code dosen’t contain integration tests , so we cannot determine if
it actually works without manually testing it out .

-> the code does not break any existing integrations tests , so , we can
tell that the most common loan scenarios do not break

-> The integration tests themselves have very low code coverage, so the
possibility that it might break something does exist (we have tickets on
Jira for improving code coverage and hope to make some progress on the same
in this GSOC)

-> To my knowledge, there isn’t any API breaking change introduced in the
develop branch since 1.2 release , so the community-app should continue to
work.

Apologies if my responses aren’t particularly helpful.

On Friday, March 8, 2019, Santosh Math <notifications@github.com> wrote:

> @vishwasbabu <https://github.com/vishwasbabu> , @ShruthiRajaram
> <https://github.com/ShruthiRajaram> , if this PR is merged and since it's
> UI version is not yet implemented, openMF/community-app#2997
> <https://github.com/openMF/community-app/issues/2997>
> Do you think it won't be any problem for existing functionality if I fetch
> latest code from develop branch of Community App to work with latest code
> from develop branch of Fineract?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/fineract/pull/465#issuecomment-470906245>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAhH6QtJyub7-o1tu6R_Ce8TbeJylp88ks5vUlHFgaJpZM4VnVLM>
> .
>


-- 
Sent from my iPad

edcable commented:
@vishwasbabu those replies were very helpful and clearly establishes the need for either a demo or for @santoshconflux to merge the UI code and manually test the functionality from his end:

@angelboxes 
The purpose of the demo is to demonstrate end to end the rates module work you've done since we don't have full integration test coverage in place. Santosh wanted to see a demo or manually test the code via the UI locally to ensure no regressions were introduced.

angelboxes commented:
@awasum Hi, I'm trying to assign the ticket to myself in Jira but I don't seem to find an option to do it. 


@santoshconflux This feature was designed for us as a way to configure multiple rates to a loan product, since our nominal interest rate is actually composed of two different rates, we needed an option for people that configure the loan products could choose between rates that were previously defined in the rates configuration module as the nominal interest rate used for the product. It just affects the UI as the rates chosen are added in the nominal interest rate field. 

About the demo, we'll work on it so you can test the changes by yourself and give us better feedback, we actually have a dev environment but we are not allowed to share it.

Thanks for all your  comments

vishwasbabu commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanTransactionsApiResource.java:
```diff
@@ -154,6 +157,7 @@ public String retrieveTransaction(@PathParam("loanId") @ApiParam(value = "loanId
         this.context.authenticatedUser().validateHasReadPermission(this.resourceNameForPermissions);
 
         LoanTransactionData transactionData = this.loanReadPlatformService.retrieveLoanTransaction(loanId, transactionId);
+        transactionData.setLoanChargePaidByList(this.loanChargePaidByReadPlatformService.getLoanChargesPaidByTransactionId(transactionId));
```
@angelboxes : Are these changes required as a part of the predefined Nominal interest rates functionality? I would assume that LoanChargePaid by shouldn't be related to Interest rates. I will mark all other such occurrences with the comment "here too" .

angelboxes replied:
These changes are not related to nominal interest at all, all changes that you pointed are related and used in the view transaction page. We use that data to calculate the VAT of a payment transaction. 

We calculate the VAT on the front side. It's just to show the detail of the charges paid in the transaction along with its VAT. It's not related to the rates functionality but we still included it this pull request.

I'm attaching an image of how the details are shown in the table.

![view transaction](https://user-images.githubusercontent.com/4413750/54057345-50638c80-41b8-11e9-8e4a-c25d2d1f9757.PNG)

angelboxes replied:
You tell me if do you think it should be included or not and I will remove it if necessary

vishwasbabu commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanChargePaidByData.java:
```diff
@@ -27,6 +27,17 @@
+        this.installmentNumber = installmentNumber;
+        this.chargeId = chargeId;
+        this.transactionId = transactionId;
+        this.name=name;
```
here too

vishwasbabu commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/data/LoanTransactionData.java:
```diff
@@ -340,4 +342,12 @@ public BigDecimal getInterestPortion() {
     	this.writeOffReasonOptions =writeOffReasonOptions;
     }
+
+    public Collection<LoanChargePaidByData> getLoanChargePaidByList() {
```
here too

vishwasbabu commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanChargePaidByReadPlatformService.java:
```diff
@@ -0,0 +1,27 @@
+
+public interface LoanChargePaidByReadPlatformService {
+
+    List<LoanChargePaidByData> getLoanChargesPaidByTransactionId(Long id);
```
here too

vishwasbabu commented on fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanChargePaidByReadPlatformServiceImpl.java:
```diff
@@ -0,0 +1,82 @@
+/**
```
here too

vishwasbabu commented on fineract-provider/src/main/resources/sql/migrations/core_db/V352__rates.sql:
```diff
@@ -0,0 +1,67 @@
+
+INSERT INTO `c_configuration`
+(`name`, `value`, `enabled`, `is_trap_door`, `description`) VALUES
+('vat-tax', 12, 0, 0, 'VAT tax'),
```
what does this configuration do? Is it related to rates functionality?

vishwasbabu reviewed (changes requested):
Does the pr contain changes unrelated to the stated functionality?

myrle-krantz commented:
Hey y'all as per our discussion here: https://lists.apache.org/thread.html/0cd7baec95e7b2b937aeec8c39afd6eb78d36708cb848a26e7b397af@%3Cdev.fineract.apache.org%3E

I've accepted angelboxes PR as an unmerged branch with the name FINERACT-614.  In the process, I resolved the merge conflict.  I have not resolved the findbugs warnings there.  Unfortunately github doesn't seem to automatically bring this (amazing!) conversation along for the ride, so those of you with outstanding comments will have to take them to the dev list.  Consider tagging your subject lines with [FINERACT-614] to make them easier to filter and find.

@angelboxes you should be able to continue working on your branch, but there's no need to edit your monocommit any longer.  Just add new commits for the adjustments you'd like to make, and continuously send PRs to this repo.  We will be able to accept most, perhaps even all of them now.  When you are done and everyone here is happy we can merge the branch to develop.
